tls/extensions/s2n_client_psk.c

```diff
@@ -181,19 +243,40 @@ static S2N_RESULT s2n_client_psk_recv_identity_list(struct s2n_connection *conn,
         uint32_t obfuscated_ticket_age = 0;
         GUARD_AS_RESULT(s2n_stuffer_read_uint32(wire_identities_in, &obfuscated_ticket_age));
 
-        /* TODO: Implement the callback to choose a PSK: https://github.com/awslabs/s2n/issues/2397
-         *
-         * When we don't have a callback configured to choose a PSK, we should fall back to accepting
-         * the first PSK identity that also exists in our list of supported PSKs. */
-        GUARD_RESULT(s2n_match_psk_identity(&conn->psk_params.psk_list, &identity, &conn->psk_params.chosen_psk));
+        wire_index++;
+    }
 
-        if (conn->psk_params.chosen_psk) {
-            conn->psk_params.chosen_psk_wire_index = wire_index;
-            return S2N_RESULT_OK;
-        }
+    struct s2n_chosen_psk chosen_psk  = { 0 };
+    uint8_t external_identity[ S2N_MAX_IDENTITY_SIZE ] = { 0 };
+    uint8_t external_secret[ S2N_MAX_SECRET_SIZE ] = { 0 };
+    chosen_psk.external_psk.identity = external_identity;
+    chosen_psk.external_psk.secret = external_secret;
+    chosen_psk.wire_index = S2N_MAX_WIRE_INDEX;
 
-        wire_index++;
+    if (conn->config->psk_selection_cb) {
+        conn->config->psk_selection_cb(conn, wire_identities, identities_count, &chosen_psk);
+    } else {
+        GUARD_AS_RESULT(s2n_config_choose_psk_cb(conn, s2n_choose_psk));
+        ENSURE_REF(conn->config->psk_selection_cb);
+        conn->config->psk_selection_cb(conn, wire_identities, identities_count, &chosen_psk);
     }
+
+    if (chosen_psk.wire_index < S2N_MAX_WIRE_INDEX) {
+        struct s2n_psk params_chosen_psk = { 0 };
+        GUARD_AS_RESULT(s2n_blob_init(&params_chosen_psk.identity, chosen_psk.external_psk.identity,
+                                    chosen_psk.external_psk.identity_length));
+        GUARD_AS_RESULT(s2n_blob_init(&params_chosen_psk.secret, chosen_psk.external_psk.secret,
+                                    chosen_psk.external_psk.secret_length));
+        params_chosen_psk.type = S2N_PSK_TYPE_EXTERNAL;
+        params_chosen_psk.obfuscated_ticket_age = 0;
+        params_chosen_psk.early_secret.data = NULL;
+        params_chosen_psk.early_secret.size = 0;
```

Comment:
Use [s2n_psk_init](https://github.com/awslabs/s2n/blob/main/tls/s2n_psk.c#L29-L38), this is why it exists.